crates/factor-outbound-networking/src/runtime_config/spin.rs

```diff
@@ -0,0 +1,273 @@
+use anyhow::{bail, ensure, Context};
+use serde::{Deserialize, Deserializer};
+use spin_factors::runtime_config::toml::GetTomlValue;
+use std::io;
+use std::{
+    fs,
+    path::{Path, PathBuf},
+};
+
+use super::{validate_host, TlsConfig};
+
+pub struct SpinTlsRuntimeConfig {
+    runtime_config_dir: PathBuf,
+}
+
+impl SpinTlsRuntimeConfig {
+    pub fn new(runtime_config_dir: impl Into<PathBuf>) -> Self {
+        Self {
+            runtime_config_dir: runtime_config_dir.into(),
+        }
+    }
+
+    /// Get the runtime configuration for client TLS from a TOML table.
+    ///
+    /// Expects table to be in the format:
+    /// ````toml
+    /// [[client_tls]]
+    /// component_ids = ["example-component"]
+    /// hosts = ["example.com"]
+    /// ca_use_webpki_roots = true
+    /// ca_roots_file = "path/to/roots.crt"
+    /// client_cert_file = "path/to/client.crt"
+    /// client_private_key_file = "path/to/client.key"
+    /// ```
+    pub fn config_from_table<T: GetTomlValue>(
+        &self,
+        table: &T,
+    ) -> anyhow::Result<Option<super::RuntimeConfig>> {
+        let Some(tls_configs) = self.tls_configs_from_table(table)? else {
+            return Ok(None);
+        };
+        let runtime_config = super::RuntimeConfig::new(tls_configs)?;
+        Ok(Some(runtime_config))
+    }
+
+    fn tls_configs_from_table<T: GetTomlValue>(
+        &self,
+        table: &T,
+    ) -> anyhow::Result<Option<Vec<TlsConfig>>> {
+        let Some(array) = table.get("client_tls") else {
+            return Ok(None);
+        };
+        let toml_configs: Vec<RuntimeConfigToml> = array.clone().try_into()?;
+
+        let tls_configs = toml_configs
+            .into_iter()
+            .map(|toml_config| self.load_tls_config(toml_config))
+            .collect::<anyhow::Result<Vec<_>>>()?;
```

Comment:
I think we want  a `context` here so that the `load_tls_config` errors are shown to the user in the context of parsing the runtime config toml file. Right now that might not be clear to the user.